src/function.ts

```diff
@@ -1,43 +1,94 @@
 import { CallExpr, isVariableReference } from "./expression";
 import { ASL } from "./asl";
 
 // @ts-ignore - imported for typedoc
 import type { AppsyncResolver, AppSyncVtlIntegration } from "./appsync";
-import { Integration } from "./integration";
+import { NativeFunctionDecl, isNativeFunctionDecl } from "./declaration";
+import { Construct } from "constructs";
+import { AnyFunction } from "./util";
+import { runtime } from "@pulumi/pulumi";
+import path from "path";
+import fs from "fs";
+import { Err, isErr } from "./error";
+import { IntegrationImpl, Integration } from "./integration";
+import { Lambda, EventBridge } from "aws-sdk";
 
-export function isFunction<P = any, O = any>(a: any): a is Function<P, O> {
+export function isFunction<P = any, O = any>(a: any): a is IFunction<P, O> {
   return a?.kind === "Function";
 }
 
 export type AnyLambda = Function<any, any>;
 
-/**
- * Wraps an {@link aws_lambda.Function} with a type-safe interface that can be
- * called from within an {@link AppsyncResolver}.
- *
- * For example:
- * ```ts
- * const getPerson = new Function<string, Person>(
- *   new aws_lambda.Function(..)
- * );
- *
- * new AppsyncResolver(() => {
- *   return getPerson("value");
- * })
- * ```
- */
-export class Function<P, O> implements Integration {
+export type FunctionClosure<P, O> = (payload: P) => Promise<O>;
+
+export interface IFunction<P, O> {
+  readonly functionlessKind: typeof Function.FunctionlessType;
+  readonly kind: typeof Function.FunctionlessType;
+  readonly resource: aws_lambda.IFunction;
+
+  (...args: Parameters<ConditionalFunction<P, O>>): ReturnType<
+    ConditionalFunction<P, O>
+  >;
+}
+
+abstract class FunctionBase<P, O>
+  implements IFunction<P, O>, Integration<FunctionBase<P, O>>
+{
   readonly kind = "Function" as const;
+  readonly native: NativeIntegration<FunctionBase<P, O>>;
+  readonly functionlessKind = "Function";
+  public static readonly FunctionlessType = "Function";
 
   readonly appSyncVtl: AppSyncVtlIntegration;
 
   // @ts-ignore - this makes `F` easily available at compile time
   readonly __functionBrand: ConditionalFunction<P, O>;
 
   constructor(readonly resource: aws_lambda.IFunction) {
-    // Integration object for appsync VTL
+    const functionName = this.resource.functionName;
+
+    // Native is used when this function is called from another lambda function's serialized closure.
+    // define this here to closure the functionName without using `this`
+    this.native = {
+      /**
+       * Wire up permissions for this function to be called by the calling function
+       */
+      bootstrap: (context: Function<any, any>) => {
+        this.resource.grantInvoke(context.resource);
+      },
+      /**
+       * This method is called from the calling runtime lambda code (context) to invoke this lambda function.
+       */
+      call: async (args, prewarmContext) => {
+        const [payload] = args;
+        const lambdaClient = prewarmContext.lambda();
```

Comment:
does this prewarmContext properly tree-shake? If i only use `lambda()`, will it also serialize other client dependencies?

Reply:
It does not! Been looking at that too.

Technically any of the keys could be requested at runtime event if they are not preWarmed. (and new entries can be added.)

I'll add a task to dynamically build the contents of the PreWarmContext instead of using an object.

Because the `aws_sdk` import is used, the impact of the extra methods is pretty minimal.

```js
function __f4() {
  return (function() {
    with({ aws_sdk_1: require("aws-sdk/lib/aws.js") }) {

return function /*lambda*/() {
    return this.getOrInit("LAMBDA", () => {
      var _a;

      return new aws_sdk_1.Lambda((_a = this.clientConfigRetriever) === null || _a === void 0 ? void 0 : _a.call(this, "LAMBDA"));
    });
  };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f5() {
  return (function() {
    with({ aws_sdk_1: require("aws-sdk/lib/aws.js") }) {

return function /*eventBridge*/() {
    return this.getOrInit("EVENT_BRIDGE", () => {
      var _a;

      return new aws_sdk_1.EventBridge((_a = this.clientConfigRetriever) === null || _a === void 0 ? void 0 : _a.call(this, "EVENT_BRIDGE"));
    });
  };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
...
var __preWarmContext_proto = {};
Object.defineProperty(__f3, "prototype", { value: __preWarmContext_proto });
Object.defineProperty(__preWarmContext_proto, "constructor", { configurable: true, writable: true, value: __f3 });
Object.defineProperty(__preWarmContext_proto, "lambda", { configurable: true, writable: true, value: __f4 });
Object.defineProperty(__preWarmContext_proto, "eventBridge", { configurable: true, writable: true, value: __f5 });
Object.defineProperty(__preWarmContext_proto, "registerCustom", { configurable: true, writable: true, value: __f6 });
Object.defineProperty(__preWarmContext_proto, "custom", { configurable: true, writable: true, value: __f7 });
Object.defineProperty(__preWarmContext_proto, "getOrInit", { configurable: true, writable: true, value: __f8 });
Object.defineProperty(__preWarmContext_proto, "getOrFail", { configurable: true, writable: true, value: __f9 });
```

Reply:
Fixed this, new API doesn't serialize unless used.